File: scripts/websocket-load-test.go
package main

import (
    "context"
    "flag"
    "fmt"
    "log"
    "net/http"
    "sync"
    "sync/atomic"
    "time"
    
    "github.com/coder/websocket"
    "github.com/coder/websocket/wsjson"
    "github.com/google/uuid"
)

// Message types
type Message struct {
    ID     string      `json:"id"`
    Type   int         `json:"type"`
    Method string      `json:"method,omitempty"`
    Params interface{} `json:"params,omitempty"`
    Result interface{} `json:"result,omitempty"`
    Error  *Error      `json:"error,omitempty"`
}

type Error struct {
    Code    int         `json:"code"`
    Message string      `json:"message"`
    Data    interface{} `json:"data,omitempty"`
}

// Metrics
type Metrics struct {
    ConnectionsCreated   int64
    ConnectionsFailed    int64
    MessagesSent        int64
    MessagesReceived    int64
    Errors              int64
    TotalLatency        int64 // in microseconds
    LatencyCount        int64
    MinLatency          int64
    MaxLatency          int64
}

func (m *Metrics) RecordLatency(latency time.Duration) {
    us := latency.Microseconds()
    atomic.AddInt64(&m.TotalLatency, us)
    atomic.AddInt64(&m.LatencyCount, 1)
    
    // Update min/max
    for {
        min := atomic.LoadInt64(&m.MinLatency)
        if min == 0 || us < min {
            if atomic.CompareAndSwapInt64(&m.MinLatency, min, us) {
                break
            }
        } else {
            break
        }
    }
    
    for {
        max := atomic.LoadInt64(&m.MaxLatency)
        if us > max {
            if atomic.CompareAndSwapInt64(&m.MaxLatency, max, us) {
                break
            }
        } else {
            break
        }
    }
}

func runClient(ctx context.Context, id int, wsURL, apiKey string, messagesPerClient int, metrics *Metrics, wg *sync.WaitGroup) {
    defer wg.Done()
    
    opts := &websocket.DialOptions{
        HTTPHeader: http.Header{
            "Authorization": []string{"Bearer " + apiKey},
        },
    }
    
    conn, _, err := websocket.Dial(ctx, wsURL, opts)
    if err != nil {
        atomic.AddInt64(&metrics.ConnectionsFailed, 1)
        log.Printf("Client %d: Failed to connect: %v", id, err)
        return
    }
    defer func() {
        if err := conn.Close(websocket.StatusNormalClosure, ""); err != nil {
            log.Printf("Client %d: Error closing connection: %v", id, err)
        }
    }()
    
    atomic.AddInt64(&metrics.ConnectionsCreated, 1)
    
    // Initialize connection
    initMsg := Message{
        ID:     uuid.New().String(),
        Type:   0, // Request
        Method: "initialize",
        Params: map[string]interface{}{
            "name":    fmt.Sprintf("load-test-agent-%d", id),
            "version": "1.0.0",
        },
    }
    
    err = wsjson.Write(ctx, conn, initMsg)
    if err != nil {
        atomic.AddInt64(&metrics.Errors, 1)
        log.Printf("Client %d: Failed to send init message: %v", id, err)
        return
    }
    atomic.AddInt64(&metrics.MessagesSent, 1)
    
    var initResp Message
    err = wsjson.Read(ctx, conn, &initResp)
    if err != nil {
        atomic.AddInt64(&metrics.Errors, 1)
        log.Printf("Client %d: Failed to read init response: %v", id, err)
        return
    }
    atomic.AddInt64(&metrics.MessagesReceived, 1)
    
    if initResp.Error != nil {
        atomic.AddInt64(&metrics.Errors, 1)
        log.Printf("Client %d: Init error: %v", id, initResp.Error.Message)
        return
    }
    
    // Send messages
    for i := 0; i < messagesPerClient; i++ {
        select {
        case <-ctx.Done():
            return
        default:
        }
        
        // Vary the message type
        var msg Message
        switch i % 3 {
        case 0:
            msg = Message{
                ID:     fmt.Sprintf("client-%d-msg-%d", id, i),
                Type:   0, // Request
                Method: "tool.list",
            }
        case 1:
            msg = Message{
                ID:     fmt.Sprintf("client-%d-msg-%d", id, i),
                Type:   0, // Request
                Method: "context.create",
                Params: map[string]interface{}{
                    "name":    fmt.Sprintf("test-context-%d-%d", id, i),
                    "content": "Load test context",
                },
            }
        case 2:
            msg = Message{
                ID:     fmt.Sprintf("client-%d-msg-%d", id, i),
                Type:   0, // Request
                Method: "context.list",
                Params: map[string]interface{}{
                    "limit": 10,
                },
            }
        }
        
        start := time.Now()
        err = wsjson.Write(ctx, conn, msg)
        if err != nil {
            atomic.AddInt64(&metrics.Errors, 1)
            continue
        }
        atomic.AddInt64(&metrics.MessagesSent, 1)
        
        var resp Message
        err = wsjson.Read(ctx, conn, &resp)
        if err != nil {
            atomic.AddInt64(&metrics.Errors, 1)
            continue
        }
        atomic.AddInt64(&metrics.MessagesReceived, 1)
        
        latency := time.Since(start)
        metrics.RecordLatency(latency)
        
        if resp.Error != nil {
            atomic.AddInt64(&metrics.Errors, 1)
        }
        
        // Small delay between messages
        time.Sleep(100 * time.Millisecond)
    }
}

func main() {
    var (
        wsURL        = flag.String("url", "ws://localhost:8080/ws", "WebSocket URL")
        apiKey       = flag.String("apikey", "test-key-admin", "API key")
        numClients   = flag.Int("connections", 10, "Number of concurrent connections")
        messagesPerClient = flag.Int("messages", 100, "Messages per connection")
        duration     = flag.Duration("duration", 60*time.Second, "Test duration")
    )
    flag.Parse()
    
    fmt.Println("WebSocket Load Test")
    fmt.Printf("URL: %s\n", *wsURL)
    fmt.Printf("Connections: %d\n", *numClients)
    fmt.Printf("Messages per connection: %d\n", *messagesPerClient)
    fmt.Printf("Duration: %v\n", *duration)
    fmt.Println()
    
    ctx, cancel := context.WithTimeout(context.Background(), *duration)
    defer cancel()
    
    metrics := &Metrics{}
    start := time.Now()
    
    var wg sync.WaitGroup
    
    // Start clients with some delay to avoid thundering herd
    for i := 0; i < *numClients; i++ {
        wg.Add(1)
        go runClient(ctx, i, *wsURL, *apiKey, *messagesPerClient, metrics, &wg)
        time.Sleep(50 * time.Millisecond) // Stagger connections
    }
    
    // Wait for all clients to finish
    wg.Wait()
    
    elapsed := time.Since(start)
    
    // Print results
    fmt.Println("\n=== Load Test Results ===")
    fmt.Printf("Duration: %v\n", elapsed)
    fmt.Printf("\nConnection Metrics:\n")
    fmt.Printf("  Created: %d\n", atomic.LoadInt64(&metrics.ConnectionsCreated))
    fmt.Printf("  Failed: %d\n", atomic.LoadInt64(&metrics.ConnectionsFailed))
    
    fmt.Printf("\nMessage Metrics:\n")
    fmt.Printf("  Sent: %d\n", atomic.LoadInt64(&metrics.MessagesSent))
    fmt.Printf("  Received: %d\n", atomic.LoadInt64(&metrics.MessagesReceived))
    fmt.Printf("  Errors: %d\n", atomic.LoadInt64(&metrics.Errors))
    
    sent := atomic.LoadInt64(&metrics.MessagesSent)
    if sent > 0 {
        fmt.Printf("  Success Rate: %.2f%%\n", float64(sent-atomic.LoadInt64(&metrics.Errors))/float64(sent)*100)
    }
    
    fmt.Printf("\nLatency Metrics:\n")
    count := atomic.LoadInt64(&metrics.LatencyCount)
    if count > 0 {
        avgLatency := atomic.LoadInt64(&metrics.TotalLatency) / count
        fmt.Printf("  Average: %.2fms\n", float64(avgLatency)/1000)
        fmt.Printf("  Min: %.2fms\n", float64(atomic.LoadInt64(&metrics.MinLatency))/1000)
        fmt.Printf("  Max: %.2fms\n", float64(atomic.LoadInt64(&metrics.MaxLatency))/1000)
    }
    
    fmt.Printf("\nThroughput:\n")
    fmt.Printf("  Messages/sec: %.2f\n", float64(sent)/elapsed.Seconds())
    
    // Exit with error if too many failures
    errorRate := float64(atomic.LoadInt64(&metrics.Errors)) / float64(sent)
    if errorRate > 0.1 { // More than 10% errors
        fmt.Printf("\nERROR: High error rate: %.2f%%\n", errorRate*100)
        log.Fatal("Load test failed due to high error rate")
    }
}

File: scripts/websocket-load-test.js
import ws from 'k6/ws';
import { check, sleep } from 'k6';
import { Counter, Rate, Trend } from 'k6/metrics';
import { uuidv4 } from 'https://jslib.k6.io/k6-utils/1.4.0/index.js';

// Custom metrics
const messagesSent = new Counter('websocket_messages_sent');
const messagesReceived = new Counter('websocket_messages_received');
const messageErrors = new Counter('websocket_message_errors');
const connectionErrors = new Counter('websocket_connection_errors');
const messageLatency = new Trend('websocket_message_latency');
const successRate = new Rate('websocket_success_rate');

// Test configuration
export const options = {
  stages: [
    { duration: '30s', target: 10 },   // Ramp up to 10 connections
    { duration: '1m', target: 50 },    // Ramp up to 50 connections
    { duration: '2m', target: 100 },   // Ramp up to 100 connections
    { duration: '1m', target: 100 },   // Stay at 100 connections
    { duration: '30s', target: 0 },    // Ramp down to 0
  ],
  thresholds: {
    'websocket_success_rate': ['rate>0.95'],
    'websocket_message_latency': ['p(95)<500', 'p(99)<1000'],
    'websocket_connection_errors': ['count<10'],
  },
};

// Test data
const BASE_URL = __ENV.MCP_SERVER_URL || 'http://localhost:8080';
const WS_URL = BASE_URL.replace('http://', 'ws://').replace('https://', 'wss://') + '/ws';
const API_KEY = __ENV.TEST_API_KEY || 'test-key-admin';

export default function () {
  const url = WS_URL;
  const params = {
    headers: {
      'Authorization': `Bearer ${API_KEY}`,
    },
    tags: { name: 'websocket_load_test' },
  };

  const res = ws.connect(url, params, function (socket) {
    socket.on('open', () => {
      console.log('WebSocket connection opened');

      // Send initialize message
      const initMsg = {
        id: uuidv4(),
        type: 0, // Request
        method: 'initialize',
        params: {
          name: `load-test-agent-${__VU}`,
          version: '1.0.0',
        },
      };
      
      socket.send(JSON.stringify(initMsg));
      messagesSent.add(1);
    });

    socket.on('message', (data) => {
      messagesReceived.add(1);
      
      try {
        const msg = JSON.parse(data);
        
        // Check for errors
        if (msg.error) {
          messageErrors.add(1);
          successRate.add(false);
          console.error(`Error response: ${msg.error.message}`);
          return;
        }
        
        successRate.add(true);
        
        // After initialization, send regular messages
        if (msg.id && msg.type === 1) { // Response
          // Send tool.list request
          const toolListMsg = {
            id: uuidv4(),
            type: 0, // Request
            method: 'tool.list',
          };
          
          const startTime = Date.now();
          socket.send(JSON.stringify(toolListMsg));
          messagesSent.add(1);
          
          // Track latency when we get the response
          socket.on('message', (responseData) => {
            try {
              const response = JSON.parse(responseData);
              if (response.id === toolListMsg.id) {
                const latency = Date.now() - startTime;
                messageLatency.add(latency);
              }
            } catch (e) {
              messageErrors.add(1);
            }
          });
          
          // Send context operations
          if (Math.random() < 0.3) { // 30% chance
            const contextMsg = {
              id: uuidv4(),
              type: 0, // Request
              method: 'context.create',
              params: {
                name: `test-context-${__VU}-${Date.now()}`,
                content: 'Load test context content',
              },
            };
            
            socket.send(JSON.stringify(contextMsg));
            messagesSent.add(1);
          }
        }
      } catch (e) {
        messageErrors.add(1);
        console.error(`Failed to parse message: ${e}`);
      }
    });

    socket.on('close', () => {
      console.log('WebSocket connection closed');
    });

    socket.on('error', (e) => {
      connectionErrors.add(1);
      console.error(`WebSocket error: ${e}`);
    });

    // Keep connection open and send periodic messages
    socket.setInterval(() => {
      const pingMsg = {
        id: uuidv4(),
        type: 0, // Request
        method: 'tool.list',
      };
      
      socket.send(JSON.stringify(pingMsg));
      messagesSent.add(1);
    }, 5000); // Every 5 seconds

    // Let the connection run for a while
    socket.setTimeout(() => {
      socket.close();
    }, 60000); // 60 seconds
  });

  check(res, {
    'WebSocket connection established': (r) => r && r.status === 101,
  });

  // Sleep to avoid overwhelming the server
  sleep(1);
}

export function handleSummary(data) {
  return {
    'stdout': textSummary(data, { indent: ' ', enableColors: true }),
    'websocket-load-test-results.json': JSON.stringify(data),
  };
}

// Helper function for text summary
function textSummary(data, options) {
  const { metrics } = data;
  let summary = '\n=== WebSocket Load Test Results ===\n\n';
  
  // Connection metrics
  summary += 'Connection Metrics:\n';
  summary += `  Total Connections: ${metrics.websocket_connection_errors ? metrics.websocket_connection_errors.values.count : 0}\n`;
  summary += `  Connection Errors: ${metrics.websocket_connection_errors ? metrics.websocket_connection_errors.values.count : 0}\n\n`;
  
  // Message metrics
  summary += 'Message Metrics:\n';
  summary += `  Messages Sent: ${metrics.websocket_messages_sent ? metrics.websocket_messages_sent.values.count : 0}\n`;
  summary += `  Messages Received: ${metrics.websocket_messages_received ? metrics.websocket_messages_received.values.count : 0}\n`;
  summary += `  Message Errors: ${metrics.websocket_message_errors ? metrics.websocket_message_errors.values.count : 0}\n`;
  summary += `  Success Rate: ${metrics.websocket_success_rate ? (metrics.websocket_success_rate.values.rate * 100).toFixed(2) : 0}%\n\n`;
  
  // Latency metrics
  if (metrics.websocket_message_latency) {
    summary += 'Latency Metrics (ms):\n';
    summary += `  Average: ${metrics.websocket_message_latency.values.avg.toFixed(2)}\n`;
    summary += `  Min: ${metrics.websocket_message_latency.values.min.toFixed(2)}\n`;
    summary += `  Max: ${metrics.websocket_message_latency.values.max.toFixed(2)}\n`;
    summary += `  P95: ${metrics.websocket_message_latency.values['p(95)'].toFixed(2)}\n`;
    summary += `  P99: ${metrics.websocket_message_latency.values['p(99)'].toFixed(2)}\n`;
  }
  
  return summary;
}